### PR TITLE
Fix settings schema missing extension-provided LSP settings on remote

### PR DESCRIPTION
**Pull request by vancez**
## Objective

Fixes an issue where language servers provided by extensions are not included in the settings schemas served by remote servers. When connecting to a remote server via the `open_remote_project_with_existing_connection` flow (dev servers, multi-workspace reopen, git worktree picker), setup-N servers never received any extension sync, so their schemas omitted extension-provided language servers, breaking language support for remote development.

## Solution

Three changes ensure that language servers provided by extensions are included in the settings schemas served by remote servers:

1. **Register connections opened via `open_remote_project_with_existing_connection` with the extension store**: dev servers, multi-workspace reopen, and the git worktree picker now register their connections with the extension store so that extensions are synced to those servers. Previously only the `open_remote_project` flow did this, so setup-N servers never received any extension sync and their schemas omitted extension language servers.

2. **Emit `ExtensionsInstalledChanged` from `HeadlessExtensionStore` after sync and install complete**: this allows the remote process to invalidate its cached settings schemas. The local `ExtensionStore` already emitted this event; without it, the remote schema cache stayed stale for the lifetime of the server process.

3. **Include available LSP adapters in the project settings schema**: this matches the user settings schema (this was fixed for `settings` in #46766 but never applied to `project_settings`).

## Testing

- Manually verified that after connecting to a remote server via the `open_remote_project_with_existing_connection` flow (including dev servers and the git worktree picker), extension language servers appear in the schema served by the remote server.
- Verified that `ExtensionsInstalledChanged` is emitted after sync and install complete, and that the remote schema cache is correctly invalidated and rebuilt.
- Verified that the `project_settings` schema includes the available LSP adapters, matching the behavior of the user settings schema.
- Reviewers should focus on the multi-workspace reopen and git worktree picker flows (the two previously unregistered flows) to confirm that extension language servers are correctly synced and appear in the schema.

## Self-Review Checklist:

- [x] I've reviewed my own diff for quality, security, and reliability
- [ ] Unsafe blocks (if any) have justifying comments
- [ ] The content adheres to Zed's UI standards ([UX/UI](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#uiux-checklist) and [icon](https://github.com/zed-industries/zed/blob/main/crates/icons/README.md) guidelines)
- [x] Tests cover the new/changed behavior
- [x] Performance impact has been considered and is acceptable

## Showcase
**Server / Project** — before the fix
image
**Local**
image

---

Release Notes:

- Fixed an issue where language servers provided by extensions were missing from settings schemas served by remote servers.

### CODE REVIEW

**Comment by vancez** [comment]
@cla-bot check

File: crates/extension_host/src/headless_host.rs [hunk 1]
```diff
@@ -103,6 +106,15 @@ impl HeadlessExtensionStore {
                 }
             }
 
+            if extensions_changed {
+                // Notify listeners (e.g. the JSON schema store) that the set of
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/headless_host.rs]: Please remove this and all other LLM-esque comments: describing verbatim a few lines below is noise.

File: crates/json_schema_store/src/json_schema_store.rs [hunk 2]
```diff
@@ -374,11 +374,26 @@ async fn resolve_dynamic_schema(
...
                 .into_iter()
-                .map(|adapter| adapter.name().to_string())
-                .collect::<Vec<_>>();
+                .map(|adapter| adapter.name())
+                .chain(languages.available_lsp_adapter_names())
+                .map(|name| name.to_string())
+                .collect();
+
+            let mut i = 0;
+            while i < lsp_adapter_names.len() {
```
Comment (SomeoneToIgnore) [inline crates/json_schema_store/src/json_schema_store.rs]: This is a relatively odd code for Rust: there are `.sort()` and `.dedup()` methods we can use instead, which is both way less code and more performant, asymptotically.

Or, collect it straight to `BTreeSet<String>`?

I see that it was copy-pasted from another line here, can we fix that too?

File: crates/workspace/Cargo.toml [hunk 3]
```diff
@@ -37,6 +37,7 @@ collections.workspace = true
 component.workspace = true
 db.workspace = true
 dirs.workspace = true
+extension_host.workspace = true
```
Comment (SomeoneToIgnore) [inline crates/workspace/Cargo.toml]: This is a no-go: `workspace` is a very fundamental crate and this blows up the crate graph.

Everyone in this chain depends in `remote`, including `project` — we can add some `Global` thing like `OnRemoteClientCreated(Rc<dyn Fn(Entity<RemoteClient>, &mut App)>)`, set it in `extension_host::init` and check for that in `Project::remote` it seems?
`recent_projects::open_remote_project` can remove the registration.

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
Thank you, but we cannot change the dependencies of a fundamental crate that easy.

I think there's a way out, posted that and a few style notes.

File: crates/extension_host/src/headless_host.rs [hunk 4]
```diff
@@ -103,6 +106,11 @@ impl HeadlessExtensionStore {
                 }
             }
 
+            if extensions_changed {
+                this.update(cx, |_, cx| notify_extensions_changed(cx))
+                    .log_err();
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/headless_host.rs]:
```suggestion
                    .ok();
```

Let's do that around, as the only way this `update` may fail is when we have no window at this point — pretty much the end of the application's lifecycle, so we do not need those logs at all.

Comment (SomeoneToIgnore) [reply crates/extension_host/src/headless_host.rs]: If we could use better names instead of `this`, would be also great, as it's more readable this way.

File: crates/extension_host/src/extension_host.rs [hunk 5]
```diff
@@ -292,6 +293,12 @@ pub fn init(
     });
 
     cx.set_global(GlobalExtensionStore(store));
+
+    cx.set_global(OnRemoteClientCreated(Rc::new(|client, cx| {
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/extension_host.rs]: Same on main, but wonder if worth fixing while we are here: after an SSH reconnect that restarts the remote server process, `HeadlessExtensionStore` starts empty, and nothing re-syncs until the local extension index next changes.
We could re-poke `ssh_registered_tx` on a reconnect event?

File: crates/project/src/project.rs [hunk 6]
```diff
@@ -1685,6 +1687,13 @@ impl Project {
             GitStore::init(&remote_proto);
             AgentServerStore::init_remote(&remote_proto);
 
+            if let Some(on_remote_client_created) = cx
+                .try_global::<OnRemoteClientCreated>()
+                .map(|on_remote_client_created| on_remote_client_created.0.clone())
+            {
+                (on_remote_client_created)(remote.clone(), cx);
```
Comment (SomeoneToIgnore) [inline crates/project/src/project.rs]: Every new registration re-syncs all registered clients, not only the new one.
With K open remote projects, opening one more costs K `SyncExtensions` round-trips, where all but one return an empty `missing_extensions` list.
K is bounded by the number of open remote projects, so not that bad, but still, can we make it better?

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, this looks almost ready.

Main concern is that

Image

definitely not there yet, can we add one, otherwise it may regress in the future?

File: crates/extension_host/src/extension_host.rs [hunk 7]
```diff
@@ -1959,13 +1966,22 @@ impl ExtensionStore {
...
+            }
+        })
+        .detach();
+
+        cx.spawn(async move |this, cx| {
+            Self::sync_extensions_to_remotes(&this, client.downgrade(), cx)
+                .await
+                .log_err();
+        })
+        .detach();
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/extension_host.rs]: Before this PR, every sync went through `ssh_registered_tx` into the single loop task in `ExtensionStore::new`, so syncs to one server were serialized.

A detached task here is unordered relative to that loop, so a registration sync can race a debounce-driven `update_remote_clients` pass to the same client (window: opening a remote project within `RELOAD_DEBOUNCE_DURATION` of an extension index change).
Both would see the same `missing_extensions`, so the worst case per collision is a duplicate upload plus the losing `install_extension` uninstalling and reinstalling the extension the winner just loaded.
The detached future also captures the strong `client` handle across the whole upload, keeping the `RemoteClient` alive after the project is closed.

I think we do not need the spawn at all: make the channel carry a payload, e.g. `UnboundedSender<Option<WeakEntity<RemoteClient>>>`, where `Some` means "sync only this client, without the debounce" and `None` keeps the current debounced full pass for reconnects?
Then this function goes back to a synchronous push + send as before the PR, the loop task stays the sole sync owner, and we still avoid the K round-trips from the other comment.

File: crates/extension_host/src/extension_host.rs [hunk 8]
```diff
@@ -1959,13 +1966,22 @@ impl ExtensionStore {
...
-        client: Entity<RemoteClient>,
-        _cx: &mut Context<Self>,
-    ) {
+    pub fn register_remote_client(&mut self, client: Entity<RemoteClient>, cx: &mut Context<Self>) {
         self.remote_clients.push(client.downgrade());
-        self.ssh_registered_tx.unbounded_send(()).ok();
+
+        cx.subscribe(&client, |this, _client, event, _cx| {
+            if matches!(event, RemoteClientEvent::Reconnected) {
+                this.ssh_registered_tx.unbounded_send(()).ok();
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/extension_host.rs]: Great, thank you for the reconnect fix: but this is also syncing all the clients now?
I think if we do the `.detach()` fix below that should go away due to `Option<WeakEntity` usage, but wanted to raise the awareness as something to be re-checked when fixed.

File: crates/remote_server/src/remote_editing_tests.rs [hunk 9]
```diff
@@ -4315,6 +4315,46 @@ async fn test_remote_trash_restore(cx: &mut TestAppContext, server_cx: &mut Test
     });
 }
 
+#[gpui::test]
+async fn test_remote_project_creation_triggers_on_remote_client_created_callback(
```
Comment (SomeoneToIgnore) [inline crates/remote_server/src/remote_editing_tests.rs]: That's a great start, thank you — do we want to cover the reconnect case too, as there's definitely some issues with that based on the other comments.

File: crates/extension_host/src/headless_host.rs [hunk 10]
```diff
@@ -85,24 +85,31 @@ impl HeadlessExtensionStore {
...
-                if let Err(e) = Self::load_extension(this.clone(), extension.clone(), cx).await {
+                if let Err(e) = Self::load_extension(store.clone(), extension.clone(), cx).await {
                     log::info!("failed to load extension: {}, {:#}", extension.id, e);
                     missing.push(extension)
                 } else if extension.dev {
                     missing.push(extension)
                 }
             }
 
+            if extensions_changed {
```
Comment (SomeoneToIgnore) [inline crates/extension_host/src/headless_host.rs:183]: If every load fails or all are missing (e.g. on the very initial load), we would notify for nothing?
Seems redundant to do so and we can check things better to exclude this case.

**Review by SomeoneToIgnore — COMMENTED** [review]
Still seeing a bunch of `this` in new subscriptions if you want to fix more along the way with the other fixes.

**Review by SomeoneToIgnore — APPROVED** [review]
Spotted that debugger parts are not synced and uninstalled at all — ended up doing a lot more changes on top.

Would you be able to test these and see if they work for you?

**Comment by vancez** [comment]
> Spotted that debugger parts are not synced and uninstalled at all — ended up doing a lot more changes on top.
> 
> Would you be able to test these and see if they work for you?

Thanks for catching the debugger sync issue and reworking this! I don't use the debugger much, but I tested your changes locally and everything checks out.
